### PR TITLE
Validation redesign part 2.

### PR DESCRIPTION
**Pull request by crunk1**
New validation:
- each step that references a resource registers itself as a creator, deleter, or user of the resource.
- registration automatically checks for generic issues, e.g. don't use something before it is created; don't use something after it is deleted; deletion must occur after ALL users; etc.
- registerUsage has an optional hook that we can use to extend validation logic on a per-resource-type basis.
- we don't need to differentiate between existing GCE URLs and Daisy-internal names anymore. Registration will treat them the same, except resources referenced by URL don't have a registered creator.

Um... I'm sure there is more. I'll clarify as questions arise.

### CODE REVIEW

File: daisy/workflow/resources.go [hunk 1]
```diff
@@ -73,17 +61,31 @@ func (rm *resourceMap) registerCreation(name string, r *resource, s *Step) error
...
-	if ok && r.creator != nil {
+	var r *resource
+	if rm.urlRgx != nil && rm.urlRgx.MatchString(name) {
+		var err error
+		r, err = rm.registerExisting(name)
+		if err != nil {
+			return err
+		}
+	} else {
+		r, _ = rm.m[name]
```
Comment (adjackura) [inline daisy/workflow/resources.go]: Drop the _, just r = rm.m[name] if you aren't worried about it not existing

File: daisy/workflow/resources.go [hunk 2]
```diff
@@ -94,27 +96,54 @@ func (rm *resourceMap) registerDeletion(name string, s *Step) error {
...
+		return nil, fmt.Errorf("partial GCE resource URL %q needs leading \"projects/PROJECT/\"", url)
+	}
+	if rm.m == nil {
+		rm.m = map[string]*resource{}
+	} else if r, ok := rm.m[url]; ok {
+		return r, nil
+	}
+	parts := strings.Split(url, "/")
+	r := &resource{real: parts[len(parts)-1], link: url, noCleanup: true}
+	rm.m[url] = r
```
Comment (adjackura) [inline daisy/workflow/resources.go]: rm.m[url] = &resource{real: parts[len(parts)-1], link: url, noCleanup: true}

File: daisy/workflow/resources.go [hunk 3]
```diff
@@ -94,27 +96,54 @@ func (rm *resourceMap) registerDeletion(name string, s *Step) error {
...
-			return fmt.Errorf("using %s %q MUST transitively depend on step %q which creates %q", rm.typeName, name, r.creator.name, name)
+	var r *resource
+	if rm.urlRgx != nil && rm.urlRgx.MatchString(name) {
+		var err error
+		r, err = rm.registerExisting(name)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		r, _ = rm.m[name]
```
Comment (adjackura) [inline daisy/workflow/resources.go]: again, no need for the _

File: daisy/workflow/step_create_disks_test.go [hunk 4]
```diff
@@ -16,10 +16,11 @@ package workflow
 
 import (
 	"context"
+	"errors"
 	"fmt"
-	"reflect"
 	"testing"
 
+	daisy_compute "github.com/GoogleCloudPlatform/compute-image-tools/daisy/compute"
```
Comment (adjackura) [inline daisy/workflow/step_create_disks_test.go]: daisyCompute

File: daisy/workflow/step_create_images_test.go [hunk 5]
```diff
@@ -16,10 +16,11 @@ package workflow
 
 import (
 	"context"
+	"errors"
 	"fmt"
-	"reflect"
 	"testing"
 
+	daisy_compute "github.com/GoogleCloudPlatform/compute-image-tools/daisy/compute"
```
Comment (adjackura) [inline daisy/workflow/step_create_images_test.go]: daisyCompute

File: daisy/workflow/step_delete_resources.go [hunk 6]
```diff
@@ -73,11 +65,7 @@ func (d *DeleteResources) run(ctx context.Context, s *Step) error {
 		wg.Add(1)
 		go func(i string) {
 			defer wg.Done()
-			r, ok := instances[w].get(i)
-			if !ok {
-				e <- fmt.Errorf("unresolved instance %q", i)
-				return
-			}
+			r, _ := instances[w].get(i)
```
Comment (adjackura) [inline daisy/workflow/step_delete_resources.go]: drop the _

File: daisy/workflow/step_delete_resources.go [hunk 7]
```diff
@@ -90,11 +78,7 @@ func (d *DeleteResources) run(ctx context.Context, s *Step) error {
 		wg.Add(1)
 		go func(i string) {
 			defer wg.Done()
-			r, ok := images[w].get(i)
-			if !ok {
-				e <- fmt.Errorf("unresolved image %q", i)
-				return
-			}
+			r, _ := images[w].get(i)
```
Comment (adjackura) [inline daisy/workflow/step_delete_resources.go]: drop the _

File: daisy/workflow/step_delete_resources.go [hunk 8]
```diff
@@ -123,11 +107,7 @@ func (d *DeleteResources) run(ctx context.Context, s *Step) error {
 		wg.Add(1)
 		go func(d string) {
 			defer wg.Done()
-			r, ok := disks[w].get(d)
-			if !ok {
-				e <- fmt.Errorf("unresolved disk %q", d)
-				return
-			}
+			r, _ := disks[w].get(d)
```
Comment (adjackura) [inline daisy/workflow/step_delete_resources.go]: drop the _

File: daisy/workflow/resources.go (hunk 1 above)
Comment (crunk1) [inline daisy/workflow/resources.go]: Done.

File: daisy/workflow/resources.go (hunk 2 above)
Comment (crunk1) [inline daisy/workflow/resources.go]: Ack.

File: daisy/workflow/resources.go (hunk 3 above)
Comment (crunk1) [inline daisy/workflow/resources.go]: Done.

File: daisy/workflow/step_create_disks_test.go (hunk 4 above)
Comment (crunk1) [inline daisy/workflow/step_create_disks_test.go]: Done.

Comment (crunk1) [inline daisy/workflow/step_create_disks_test.go]: Python made an appearance here.

File: daisy/workflow/step_create_images_test.go (hunk 5 above)
Comment (crunk1) [inline daisy/workflow/step_create_images_test.go]: Done.

File: daisy/workflow/step_delete_resources.go (hunk 6 above)
Comment (crunk1) [inline daisy/workflow/step_delete_resources.go]: Ack.

File: daisy/workflow/step_delete_resources.go (hunk 7 above)
Comment (crunk1) [inline daisy/workflow/step_delete_resources.go]: Ack.

File: daisy/workflow/step_delete_resources.go (hunk 8 above)
Comment (crunk1) [inline daisy/workflow/step_delete_resources.go]: Ack.

**Pull request merged**